### DISCUSSION TITLE
pnpm command not recognized (VS Code)

### DISCUSSION THREAD

**Discussion by sillvva** [pnpm/pnpm#4810] [Q&A]
I've installed pnpm on Windows. The command works in Windows PowerShell Terminal, but I can't get it to work in VS Code PowerShell Terminal. I've tried removing the terminal and making a new one, restarting VS Code, updating VS Code. The command is still not recognized.

```
pnpm: The term 'pnpm' is not recognized as a name of a cmdlet, function, script file, or executable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
```

I couldn't find anything in Google search, Bing Search, or Github search to resolve the issue.

**Comment by sillvva** [comment]
Nevermind. I used the standalone script install. It worked after using npm install method.

**Comment by woss** [comment]
i don't think that's a correct answer, this is just another way to install the pnpm. 

the correct answer is, that you should have added the pnpm binary to the path. and it seems you are using windows which requires you to either use the generally accepted paths which are included in the environment or manually set up the directory where you store the custom binaries that are included in the env.

**Reply by sillvva** [reply]
Where is the documentation for doing that? After running the standalone install, where are the binaries? They're not in the directory where I ran the install.

I marked it as the answer, because it resolved the issue of not being able to run the command in the VS Code terminal.

Adding the binary to the path was my first choice, but since I don't know where the binaries are, I resorted to scrolling down and looking for other install methods. In all honesty, I think the npm install method should be the default, because it comes with node.

**Reply by woss** [reply]
You can start with this [search](https://duckduckgo.com/?q=how+to+add+custom+path+to+windows&t=brave&ia=web) then read this [article](https://dev.to/noejon/setting-up-a-node-js-development-environment-in-2-4eb1) which hopefully will convince you not to use windows for development but Linux 😊

> I think the npm install method should be the default, because it comes with node.

the `pnpm` comes with the node manager, so the best thing is to use `pnpm` without the node and then install the needed version.

**Reply by zkochan** [reply]
Actually, the standalone script should automatically add the pnpm directory to the path. This script:

```
iwr https://get.pnpm.io/install.ps1 -useb | iex
```

It runs `pnpm setup` after downloading pnpm. And `pnpm setup` adds the pnpm home directory to the windows environment registry. Here is the related code: https://github.com/pnpm/components/blob/main/os/env/path-extender-windows/path-extender-windows.ts

**Reply by Selene29** [reply]
Here is an additional tip, if you still run in this error despite all of the above on Windows 10.
The variable "PNPM_HOME" with value "C:\Users\XXXX\AppData\Local\pnpm" gets added with the install script. Under "Path" it adds the value "%PNPM_HOME%" which Windows will replace with "C:\Users\XXXX\AppData\Local\pnpm" as expected.

Here is the deal:
Despite every single program (cmd & new terminal with powershell) registering the command "pnpm" and reporting:
```
PS C:\Users\XXX> pnpm
Version 7.13.2 (compiled to binary; bundled Node.js v14.19.2)
```
The powershell (not opened with new terminal) still could not find the "pnpm" command.
And consequently VS Code could also not find it.
A pc restart did not fix it.

Solution:
Replace in "Path" "%PNPM_HOME%" explicit with "C:\Users\XXXX\AppData\Local\pnpm".
Press Ok, check powershell and it should be fixed.

==> Windows is weird sometimes....

**Reply by zkochan** [reply]
This looks related: https://github.com/pnpm/pnpm/issues/4658#issuecomment-1257197509

I wonder how to fix it.

**Reply by Fred-Vatin** [reply]
> You can start with this [search](https://duckduckgo.com/?q=how+to+add+custom+path+to+windows&t=brave&ia=web) then read this [article](https://dev.to/noejon/setting-up-a-node-js-development-environment-in-2-4eb1) which hopefully will convince you not to use windows for development but Linux 😊

How to not be helpful and disgust every user to use Linux.

**Reply by Fred-Vatin** [reply]
> Actually, the standalone script should automatically add the pnpm directory to the path. This script:
> 
> ```
> iwr https://get.pnpm.io/install.ps1 -useb | iex
> ```

It does. `%PNPM_HOME%` is added to the path. It works when using pnpm in the Windows terminal but not in VSCode terminal. Using the same powershell version, it then returns : `The term 'pnpm' is not recognized as a name of a cmdlet...`

I tried to play with VSCode settings and succeeded to fix this issue :
```jsonc
"terminal.integrated.profiles.windows": {
  "PowerShell": {
	  "path": "pwsh",
	  "icon": "terminal-powershell",
  },
},
"terminal.integrated.env.windows": {
  "PATH": "${env:PATH};${workspaceFolder}\\node_modules\\.bin;${env:PNPM_HOME}"
},
```

You may need to restart VSCode after adding these settings.

**EDIT**: as the current version 1.95, I recommend to delete the `"terminal.integrated.env.windows"` settings because if `${workspaceFolder}` doesn't exist, `${env:PATH}` is not substituted. See [#212062](https://github.com/microsoft/vscode/issues/212062) which has not been fixed as it is said.

**Reply by HarshAggarwals** [reply]
Thank you so much @Fred-Vatin, your solution resolved my issue. Adding these setting in package.json, make vs recognize pnpm command

**Reply by wildrap** [reply]
It does not create folders on app/local but the pnpm are in app/roaming even with the env on it. it still does not work

**Reply by hotsu0p** [reply]
I just needed to restart vs code

**Reply by azmek** [reply]
a simple restart of vs code resolved the issue

**Comment by privatenumber** [comment]
I'm on a fresh copy of [Windows 11 Enterprise (Evaluation) via VirtualBox](https://developer.microsoft.com/en-us/windows/downloads/virtual-machines/) and I was able to work around this by copying the `PNPM_HOME` _User variable_ to _System variable_.

<img width="453" alt="Screen Shot 2022-10-12 at 5 39 32 PM" src="https://user-images.githubusercontent.com/1075694/195294266-2422e2ec-ad4d-4a4b-9e8f-3788947bda91.png">

**Reply by alex-7799** [reply]
this is what worked for me. In my case, the pnpm is in C: disk & pwd is in D: disk

**Reply by g8up** [reply]
I change the "PNPM_HOME" to lowercase "pnpm_home", and so is the refer "%pnpm_home%" in `Path`. It works.
I don't know why.

**Reply by immortal-blog** [reply]
This workes for me! Thank you.

**Reply by loeclos** [reply]
Added `%PNPM_HOME%` to the system `Path` variable and it worked!

**Reply by samedev3** [reply]
Funcionou perfeitamente, obrigado cara, que Deus te abencoe

**Reply by c4nzin** [reply]
Thanks!

**Comment by bipulmgr** [comment]
> **First, use pnpm standalone script. It will worked.**
[https://pnpm.io/installation](url)

If that didn't work then copy pnpm.exe file from

>    C:\Users\hp\AppData\Local\pnpm

folder and paste it into

  >  C:\Program Files\nodejs
    
  folder.

**Comment by ashwth** [comment]
Windows 11 here. I have used the standalone script to install it. In my case restarting all instances of vs code solved the problem.

**Reply by kevinkirsten** [reply]
worked for me too, thanks!

**Comment by EvilChan** [comment]
Replace in "Path" "%PNPM_HOME%" explicit with "%PNPM_HOME%\\".
Press Ok, check powershell and it should be fixed.

**Comment by teocto** [comment]
@Fred-Vatin thank you !!!

**Comment by marcus-evans** [comment]
Here's the proposed solution on handling this issue

1. First check the execution policy for the current user .
Please copy the code below and execute this on your vscode terminal
```
Get-ExecutionPolicy -Scope CurrentUser
```

2. if the result is `Undefined` change the execution policy for current user.
Please copy the code below and execute this on your vscode terminal
```
Set-ExecutionPolicy -ExecutionPolicy RemoteSigned -Scope CurrentUser
```

3. Then check if the execution policy are change to `RemoteSigned`
 ```
Get-ExecutionPolicy -Scope CurrentUser
```
If the execution is successfully changed from `Undefine` to `RemoteSigned` then you can now test your any `pnpm` command directly to your vscode terminal.
Sample pnpm command
```
pnpm --version
```

Resources: 
Microsoft: [about_Execution_Policies](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_execution_policies?view=powershell-7.3)

**Reply by aix216** [reply]
i don't have administrator privileges.
It workes for me! Thank you very much.

**Comment by LouieGuides** [comment]
In my case, I open the powershell as admin and paste this
iwr https://get.pnpm.io/install.ps1 -useb | iex

Then Install the LTS version of Node.js
pnpm env use --global lts

Then go to the specific folder by using cd command and check node version if it was their
node -v

Done!

**Comment by zongren** [comment]
In my case, after changing vscode to run as administrator, the issue solved.

**Comment by Israr-11** [comment]
I solved the problem by opening the Power Shell as 'Administrator' and running the standalone script from pnpm's documentation. 

_iwr https://get.pnpm.io/install.ps1 -useb | iex_

**Reply by arshad-yaseen** [reply]
This works for me ❤️

**Comment by iBz-04** [comment]
right click VSCODE and choose "run as administrator" that's all

**Reply by serega404** [reply]
It bad way

**Comment by sharabhh** [comment]
So I went with @privatenumber method (thanks!), Go to the location in your C drive
![image](https://github.com/user-attachments/assets/e145ed29-7357-4c7a-90b5-f95cf1afec43)

and add the path to your system environment path variables.
Re launch VS code. It worked for me at least <3

**Comment by Zoly90** [comment]
Windows user, installed it through PowerShell, this worked for me. Untill than the pnpm folder was completly empty. So for PowerShell use this command:
`Invoke-WebRequest https://get.pnpm.io/install.ps1 -UseBasicParsing | Invoke-Expression`

**Comment by xyfll7** [comment]
change the file `C:\Users\NAME\AppData\Local\pnpm\pnpm`
to `C:\Users\NAME\AppData\Local\pnpm\pnpm.exe`
then is works

**Comment by ODtian** [comment]
After looking into it, the problem seems to be caused by the order variables are processed, which depends on their names. Because `PATH` comes before `PNPM_HOME` alphabetically, the system figures out `PATH` first.

The trouble is, your `PATH` setting uses `%PNPM_HOME%` to point to the pnpm directory. But when the system looks at `PATH`, it hasn't figured out what `PNPM_HOME` actually means yet. 

So, instead of putting the real pnpm path into `PATH`, it just leaves the text `%PNPM_HOME%` in there. That's not a real path, so the pnpm command can't be found.

**Easy Fix:** Just copy the actual path stored in your `PNPM_HOME` variable and paste it directly into your `PATH` variable, replacing the `%PNPM_HOME%` part.
